hazelcast/test/src/compact_test.cpp

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#ifdef HZ_BUILD_WITH_SSL
+#include <openssl/crypto.h>
+#endif
+
+#include <hazelcast/client/lifecycle_event.h>
+#include <hazelcast/client/client_config.h>
+#include <hazelcast/client/connection/ClientConnectionManagerImpl.h>
+#include <hazelcast/client/connection/Connection.h>
+#include <hazelcast/client/exception/protocol_exceptions.h>
+#include <hazelcast/client/hazelcast_client.h>
+#include <hazelcast/client/hazelcast_json_value.h>
+#include <hazelcast/client/imap.h>
+#include <hazelcast/client/impl/Partition.h>
+#include <hazelcast/client/initial_membership_event.h>
+#include <hazelcast/client/internal/nearcache/impl/NearCacheRecordStore.h>
+#include <hazelcast/client/internal/nearcache/impl/store/NearCacheDataRecordStore.h>
+#include <hazelcast/client/internal/nearcache/impl/store/NearCacheObjectRecordStore.h>
+#include <hazelcast/client/internal/socket/SSLSocket.h>
+#include <hazelcast/client/lifecycle_listener.h>
+#include <hazelcast/client/pipelining.h>
+#include <hazelcast/client/serialization_config.h>
+#include <hazelcast/client/serialization/pimpl/data_input.h>
+#include <hazelcast/client/serialization/serialization.h>
+#include <hazelcast/util/AddressHelper.h>
+#include <hazelcast/util/concurrent/locks/LockSupport.h>
+#include <hazelcast/util/MurmurHash3.h>
+#include <hazelcast/util/Util.h>
+
+#include "ClientTest.h"
+#include "HazelcastServer.h"
+#include "HazelcastServerFactory.h"
+#include "TestHelperFunctions.h"
+#include "serialization/Serializables.h"
+#include "remote_controller_client.h"
+
+#if defined(WIN32) || defined(_WIN32) || defined(WIN64) || defined(_WIN64)
+#pragma warning(push)
+#pragma warning(disable : 4996) // for unsafe getenv
+#endif
+using namespace hazelcast::client::serialization::pimpl;
+namespace hazelcast {
+namespace client {
+namespace compact {
+namespace test {
+
+struct main_dto
+{
+    main_dto()
+      : i(0)
+      , str("")
+    {}
+
+    explicit main_dto(int i, std::string str)
+      : i(i)
+      , str(std::move(str))
+    {}
+
+    bool operator==(const main_dto& rhs) const
+    {
+        if (i != rhs.i) {
+            return false;
+        }
+        return str == rhs.str;
+    }
+
+    int i;
+    std::string str;
+};
```

Comment:
```suggestion
struct main_dto
{
    int i;
    std::string str;
};
```

Reply:
Did you know this before https://stackoverflow.com/a/5740343 ?
I will use your suggestion and implement equality outside of the struct.

Reply:
Fixed here 47e3cfe18994d9262289179422d6143ae23cc209